sdk/go/x/auto/local_workspace.go

```diff
@@ -0,0 +1,719 @@
+	}
+
+	if lwOpts.Repo != nil {
+		// now do the git clone
+		projDir, err := setupGitRepo(ctx, workDir, lwOpts.Repo)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace, unable to enlist in git repo")
+		}
+		workDir = projDir
+	}
+
+	var program pulumi.RunFunc
+	if lwOpts.Program != nil {
+		program = lwOpts.Program
+	}
+
+	var pulumiHome *string
+	if lwOpts.PulumiHome != nil {
+		pulumiHome = lwOpts.PulumiHome
+	}
+
+	l := &LocalWorkspace{
+		workDir:    workDir,
+		program:    program,
+		pulumiHome: pulumiHome,
+	}
+
+	if lwOpts.Project != nil {
+		err := l.WriteProjectSettings(ctx, lwOpts.Project)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace, unable to save project settings")
+		}
+	}
+
+	for fqsn := range lwOpts.Stacks {
+		s := lwOpts.Stacks[fqsn]
+		err := l.WriteStackSettings(ctx, fqsn, &s)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace")
+		}
+	}
+
+	return l, nil
+}
+
+type localWorkspaceOptions struct {
+	// WorkDir is the directory to execute commands from and store state.
+	// Defaults to a tmp dir.
+	WorkDir string
+	// Program is the Pulumi Program to execute. If none is supplied,
+	// the program identified in $WORKDIR/pulumi.yaml will be used instead.
+	Program pulumi.RunFunc
+	// PulumiHome overrides the metadata directory for pulumi commands
+	PulumiHome *string
+	// Project is the project settings for the workspace
+	Project *workspace.Project
+	// Stacks is a map of [fqsn -> stack settings objects] to seed the workspace
+	Stacks map[string]workspace.ProjectStack
+	// Repo is a git repo with a Pulumi Project to clone into the WorkDir.
+	Repo *GitRepo
+}
+
+// LocalWorkspaceOption is used to customize and configure a LocalWorkspace at initialization time.
+// See Workdir, Program, PulumiHome, Project, Stacks, and Repo for concrete options.
+type LocalWorkspaceOption interface {
+	applyLocalWorkspaceOption(*localWorkspaceOptions)
+}
+
+type localWorkspaceOption func(*localWorkspaceOptions)
+
+func (o localWorkspaceOption) applyLocalWorkspaceOption(opts *localWorkspaceOptions) {
+	o(opts)
+}
+
+// GitRepo contains info to acquire and setup a Pulumi program from a git repository.
+type GitRepo struct {
+	// URL to clone git repo
+	URL string
+	// Optional path relative to the repo root specifying location of the pulumi program.
+	// Specifying this option will update the Worspace's WorkDir accordingly.
+	ProjectPath string
+	// Optional branch to checkout
+	Branch string
+	// Optional commit to checkout
+	CommitHash string
+	// Optional function to execute after enlisting in the specified repo.
+	Setup SetupFn
+}
+
+// SetupFn is a function to execute after enlisting in a git repo.
+// It is called with a PATH containing the pulumi program post-enlistment.
+type SetupFn func(context.Context, string) error
+
+// WorkDir is the directory to execute commands from and store state.
+func WorkDir(workDir string) LocalWorkspaceOption {
```

Comment:
Not sure if it was obvious, but `GitRepo.Setup` is a callback that gets executed post enlistment with the `WorkDir`. Was that not a useful helper for your use case?

Reply:
Ah neat, I missed that the first go around.

Reply:
Tried to move to using the `GitRepo.Setup` callback, but for the use case of installing deps, I also need to load the project to get its runtime for handling the respective package manager. 

The current def of `type SetupFn func(context.Context, string) error` is limiting in what args can be passed in without duplicating workspace related functions e.g. create and select, and then loading a proj to extract it's runtime.

Am I going about this wrong?

Reply:
I think we should take a look through the code your trying to integrate with on Monday. I think this callback should be changed so that it's called with the full Workspace, rather than just the WorkDir. At this point the stack is already created/selected, but you'll have the utilities on the Workspace object to load the project file, etc if needed.